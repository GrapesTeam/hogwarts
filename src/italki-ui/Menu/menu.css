.menu {
  position: relative;
  display: inline-block;
  margin: 0;
  padding: 0;
  height: 42px;
  font-family: 'Roboto';
  font-size: 14px;
  border: solid 1px #cccccc;
  cursor: pointer;
  user-select: none;
}

.menu-hide {
  overflow: hidden;
  cursor: pointer;
}

.menu-title {
  padding: 13px 28px 13px 12px;
}

.menu-items {
  position: absolute;
  top: 100%;
  left: -1px;
  width: 100%;
  max-height: 300px;
  overflow-y: scroll;
  border: solid 1px #cccccc;
  transition: max-height 0.3s;
}

.menu-hide .menu-items {
  max-height: 0;
}

.menu-item {
  padding: 13px 9px;
  height: 16px;
}

.menu-item:hover {
  background: rgba(204, 204, 204, 0.2);
  cursor: pointer;
}

.arrow-down, .arrow-up {
  position: absolute;
  display: inline-block;
  top: 15px;
  right: 8px;
  width: 16px;
  height: 16px;
  transition: transform 0.2s;
}

.arrow-down {
  transform: rotate(45deg);
}

.arrow-up {
  transform: translate(1px, -2px) rotate(-135deg);
}

.arrow-up:before, .arrow-down:before {
  content: "";
  position: absolute;
  display: inline-block;
  left: 4px;
  top: 7px;
  width: 6px;
  height: 2px;
  background-color: #333333;
}

.arrow-up:after, .arrow-down:after {
  content: "";
  position: absolute;
  display: inline-block;
  top: 5px;
  left: 6px;
  width: 6px;
  height: 2px;
  background-color: #333333;
  transform: rotate(90deg);
}

.checkmark {
  position: relative;
  display: inline-block;
  margin-right: 7px;
}
.checkmark:after {
  content: "";
  display: block;
  width: 4px;
  height: 8px;
  border: solid #ff4338;
  border-width: 0 2px 2px 0;
  transform: rotate(45deg);
}