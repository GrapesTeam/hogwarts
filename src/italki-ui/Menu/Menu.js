import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import cx from 'classnames';
import MenuItem from './MenuItem';
import './menu.css';

export const MenuContext = React.createContext({
  label: '',
  selected: -1,
});

class Menu extends Component {
  static defaultProps = {
    search: false,
    placeholder: 'Please select',
  };

  state = {
    selected: this.props.selected || -1,
    open: false,
    value: '',
    label: '',
    searchText: '',
  };

  componentWillMount() {
    document.addEventListener('click', this.closeMenu);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.closeMenu);
  }

  closeMenu = event => {
    if (this.state.open) {
      if (!ReactDOM.findDOMNode(this).contains(event.target)) {
        if (this.props.search) {
          this.search.blur();
        }
        this.setState({
          ...this.state,
          open: false,
        });
      }
    }
  };

  openMenu = () => {
    if (this.props.search) {
      this.search.focus();
      this.search.setSelectionRange(
        this.search.value.length * 2,
        this.search.value.length * 2
      );
    }
    this.setState({
      ...this.state,
      open: true,
      searchText: '',
    });
  };

  setActive = actived => {
    this.placeholder = actived.label;
    this.setState({
      open: false,
      selected: actived.index,
      value: actived.value,
      label: actived.label,
    });
  };

  handleInputChange = () => {
    this.setState({
      ...this.state,
      searchText: this.search.value,
    });
  };

  renderInput() {
    const placeholder = this.placeholder || this.props.placeholder;
    return (
      <input
        placeholder={placeholder}
        type="text"
        ref={search => (this.search = search)}
        onChange={this.handleInputChange}
      />
    );
  }

  renderChoice() {
    const { children, placeholder } = this.props;
    let choice;
    const childrenArr = React.Children.toArray(children);
    if (/^\d+G\d+$/.test(this.state.selected)) {
      const groupIndex = this.state.selected.split('G')[0];
      const choiceIndex = this.state.selected.split('G')[1];
      choice =
        childrenArr[groupIndex - 1] &&
        childrenArr[groupIndex - 1].props.children[choiceIndex - 1] &&
        childrenArr[groupIndex - 1].props.children[choiceIndex - 1].props
          .children;
    }
    if (typeof this.state.selected === 'number') {
      choice =
        childrenArr[this.state.selected - 1] &&
        childrenArr[this.state.selected - 1].props.children;
    }
    choice = choice || placeholder;
    return choice;
  }

  checkChildren = child => {
    if (child.props.children && typeof child.props.children !== 'string') {
      return React.Children.forEach(child.props.children, this.checkChildren);
    }
    return (
      child.props.children &&
      child.props.children.indexOf(this.state.searchText) > -1
    );
  };

  flatterned = [];
  flattenChildren = children => {
    React.Children.forEach(children, child => {
      if (child.props.children && typeof child.props.children !== 'string') {
        return React.Children.forEach(
          child.props.children,
          this.flattenChildren
        );
      }
      this.flatterned.push(child.props.children);
    });
    return this.flatterned;
  };

  renderOptions() {
    const { children } = this.props;
    const { searchText } = this.state;

    let noResults = true;
    const flattenedChildren = this.flattenChildren(this.props.children);
    for (let i = 0; i < flattenedChildren.length; i++) {
      if (
        flattenedChildren[i] &&
        flattenedChildren[i].indexOf(this.state.searchText) > -1
      ) {
        noResults = false;
        break;
      }
    }
    const regx = new RegExp(searchText);
    return noResults ? (
      <MenuItem disabled>No Results</MenuItem>
    ) : (
      React.Children.map(
        children,
        (item, key) =>
          item.type.name === 'EnhancedMenuGroup' ||
          regx.test(flattenedChildren[key])
            ? React.cloneElement(item, {
                selected: this.state.selected,
                index: key + 1,
                setActive: this.setActive,
                searchText,
              })
            : null
      )
    );
  }

  render() {
    const { classname, search, style } = this.props;
    const { label, open, selected } = this.state;
    const classnames = cx('menu', classname, {
      'menu-hide': !open,
    });
    const arrowClassNames = cx({
      'arrow-down': !open,
      'arrow-up': open,
    });
    return (
      <div className={classnames} style={style}>
        {search ? this.renderInput() : null}
        <div onClick={this.openMenu} className="menu-title">
          {this.renderChoice()}
          <span className={arrowClassNames} />
        </div>
        <div className="menu-items">
          <MenuContext.Provider value={{ label, selected }}>
            {this.renderOptions()}
          </MenuContext.Provider>
        </div>
      </div>
    );
  }
}

export default Menu;
