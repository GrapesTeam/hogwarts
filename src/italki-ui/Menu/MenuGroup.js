// @flow
import * as React from 'react';
import { MenuContext } from './Menu';

type Props = {
  name: string,
  children: React.Node,
  index: number,
  searchText: string,
  selected: number,
  setActive: Function,
};

class MenuGroup extends React.PureComponent<Props> {
  flatterned = [];
  flattenChildren = children => {
    React.Children.forEach(children, child => {
      if (child.props.children && typeof child.props.children !== 'string') {
        return React.Children.forEach(
          child.props.children,
          this.flattenChildren
        );
      }
      this.flatterned.push(child.props.children);
    });
    return this.flatterned;
  };

  renderOptions() {
    const {
      children,
      index,
      name,
      searchText,
      selected,
      setActive,
    } = this.props;

    let isMatched = false;
    const flattenedChildren = this.flattenChildren(children);
    const regx = new RegExp(searchText);
    for (let i = 0; i < flattenedChildren.length; i++) {
      if (regx.test(flattenedChildren[i])) {
        isMatched = true;
        break;
      }
    }
    return (
      <div className="menu-group">
        {isMatched ? <h1 className="menu-group-name">{name}</h1> : null}
        {React.Children.map(
          children,
          (item, key) =>
            regx.test(flattenedChildren[key])
              ? React.cloneElement(item, {
                  actived: key === selected,
                  index: `${index}G${key + 1}`,
                  setActive,
                })
              : null
        )}
      </div>
    );
  }

  render() {
    return this.renderOptions();
  }
}

const EnhancedMenuGroup = (props: any) => (
  <MenuContext.Consumer>
    {({ label }) => <MenuGroup {...props} label={label} />}
  </MenuContext.Consumer>
);

export default EnhancedMenuGroup;
