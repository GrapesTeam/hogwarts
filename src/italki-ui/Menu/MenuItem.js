// @flow
import * as React from 'react';
import cx from 'classnames';
import { MenuContext } from './Menu';

type Props = {
  children?: React.Node,
  index: number | string,
  onClick?: Function,
  setActive: Function,
  actived: boolean,
  disabled: boolean,
  value: string,
  label: string,
};

class MenuItem extends React.PureComponent<Props> {
  static defaultProps = {
    value: '',
    disabled: false,
  };

  handleClick = () => {
    if (!this.props.disabled) {
      this.props.setActive({
        index: this.props.index,
        value: this.props.value,
        label: this.props.children,
      });
      if (this.props.onClick) {
        this.props.onClick();
      }
    }
  };

  render() {
    const { children, disabled, index } = this.props;
    const classnames = cx('menu-item', {
      'menu-item-disabled': disabled,
    });
    return (
      <MenuContext.Consumer>
        {({ selected }) => (
          <div className={classnames} onClick={this.handleClick}>
            {selected === index ? <div className="checkmark" /> : null}
            <div>{children}</div>
          </div>
        )}
      </MenuContext.Consumer>
    );
  }
}

export default MenuItem;
