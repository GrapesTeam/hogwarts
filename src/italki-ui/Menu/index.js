import MenuItem from './MenuItem';
import MenuGroup from './MenuGroup';

export { default } from './Menu';
export { MenuItem, MenuGroup };
