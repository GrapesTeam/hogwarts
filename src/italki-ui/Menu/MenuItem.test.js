import React from 'react';
import render from 'react-test-renderer';
import { mount } from 'enzyme';
import Menu, { MenuItem } from '../Menu';

describe('MenuItem', () => {
  it('renders correctly', () => {
    const tree = render.create(<MenuItem value="test">Test</MenuItem>);
    expect(tree).toMatchSnapshot();
  });

  it('should fire click event', () => {
    const click = jest.fn();
    const wrapper = mount(
      <Menu>
        <MenuItem value="test" onClick={click}>
          Test
        </MenuItem>
      </Menu>
    );
    wrapper.find('.menu-item').simulate('click');
    expect(click).toBeCalled();
  });

  it('should not fire click event if click is not existed', () => {
    const wrapper = mount(
      <Menu>
        <MenuItem value="test">Test</MenuItem>
      </Menu>
    );
    wrapper.find('.menu-item').simulate('click');
    expect(wrapper.find('.checkmark').length).toBe(1);
  });

  it('should choose item if onClick is missing', () => {
    const wrapper = mount(
      <Menu selected={0}>
        <MenuItem value="test">Test</MenuItem>
      </Menu>
    );
    expect(wrapper.find('.checkmark').length).toBe(1);
  });
});
