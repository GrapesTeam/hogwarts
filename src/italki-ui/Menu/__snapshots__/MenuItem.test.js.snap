// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem renders correctly 1`] = `
<div
  className="menu-item"
  onClick={[Function]}
>
  <div>
    Test
  </div>
</div>
`;
